### ISSUE TITLE
Relationship between Associations and Dependancies

### ISSUE THREAD

**Issue by anubh-v** [nus-cs2103-AY1819S1/forum#210]
Hi all,
I wish to clarify the relationship between associations and dependancies.
I see the following 2 possibilities:

1. Associations and dependancies are mutually exclusive.
2. Associations imply dependancies. In other words, if there is an association between 2 objects, there is also a dependancy between them. 

To illustrate, we can use this question from the practice exam.
![image](https://user-images.githubusercontent.com/35621759/49338994-77646c80-f665-11e8-980e-41024972c075.png)

The arrow from Container to Lid shows that there is an association between Container and Lid.
But is there also a dependancy between Container and Lid?

**Comment by anubh-v** [comment]
I came across this [article](https://martinfowler.com/bliki/DependencyAndAssociation.html) that says that associations imply dependancies.

> Associations also imply dependency, if there is an association between two classes, there is also a dependency. But I can't imagine a case where you would show that dependency as an extra line on the diagram. The association implies it, as does a generalization.

**Comment by anubh-v** [comment]
On the other hand, if we consider the first sentence of the following paragraph (from textbook) as a definition, then dependancies and associations are mutually exclusive.

![image](https://user-images.githubusercontent.com/35621759/49339067-deceec00-f666-11e8-8190-cca1de2d4a0b.png)

**Comment by damithc** [comment]
> 1. Associations and dependancies are mutually exclusive.
> 1. Associations imply dependancies. In other words, if there is an association between 2 objects, there is also a dependancy between them.

It's the second one.

The two things are not mutually exclusive, but we show a dependency arrow only when there is no association arrow. If there is an association arrow, it is pointless to draw a dependency arrow as well; you can think of it as a solid arrow already contains a dashed arrow.

**Comment by pengnam** [comment]
@damithc In this diagram, is there an association from Lid to Container?
I understand that arrow head indicates the navigability of an association, but I am not sure if an association hence becomes 1-way instead of 2-way.

**Comment by pengnam** [comment]
![image](https://user-images.githubusercontent.com/28092058/49424982-3a1fec00-f7d7-11e8-91e4-10faaef188ba.png)
In an association that is only navigable in one-direction, should we ever have a multiplicity on the side that is not an arrow head?

**Comment by damithc** [comment]
> @damithc In this diagram, is there an association from Lid to Container?
> I understand that arrow head indicates the navigability of an association, but I am not sure if an association hence becomes 1-way instead of 2-way.

There is an association _between_ `Lid` and `Container` and the navigability is from `Container` to `Lid`. As the navigabiity is only in one direction, it is not a bidirectional association.

**Comment by damithc** [comment]
> In an association that is only navigable in one-direction, should we ever have a multiplicity on the side that is not an arrow head?

You can have multiplicity on either end of the association irrespective of navigability as each association has two multiplicities e.g., in the `made by` association between `Car` and `Company` the two multiplicities are, 
1. How many Car objects can be made by one Company?
1. How many Compnaies could be involved in making a given Car object?

**Comment by pengnam** [comment]
@damithc 

Thank you!

https://user-images.githubusercontent.com/28092058/49424982-3a1fec00-f7d7-11e8-91e4-10faaef188ba.png

To make sure that I am on the same page:
In the case of the diagram above, a student can be part of 1 to 3 teams? Even if this multiplicity can't be enforced.

**Comment by damithc** [comment]
> To make sure that I am on the same page:
> In the case of the diagram above, a student can be part of 1 to 3 teams? Even if this multiplicity can't be enforced.

Yes, that's correct. It is not _impossible_ to enforce this multiplicity either e.g., we can write code to go through the entire object structure to ensure no student is part of more than three teams.

**Comment by pengnam** [comment]
@damithc Thank you
